### ISSUE TITLE
NullPointerException when publishing checks on unhealthy coverage

### ISSUE THREAD

**Issue by agaudreault-jive** [jenkinsci/code-coverage-api-plugin#193]
### Description 

Our build fails when it tries to report the coverage to GitHub with a status checks

#### Version report

Jenkins and plugins versions report:

```
Jenkins: 2.280
Code Coverage API Plugin: 1.3.2
Checks API: 1.6.1
```

#### Reproduction steps

Running with pipeline step

```groovy
  publishCoverage(
    adapters: [coberturaAdapter('cobertura-coverage.xml')],
    globalThresholds: [
		[
	      thresholdTarget: 'Method',
	      unhealthyThreshold: '60',
	      unstableThreshold: '40',
  		]
    ],
    failUnhealthy: true, // This setting will cause the exception and prevent the status checks from working
    failUnstable: false
  )
```

```
[Pipeline] publishCoverage
10:56:48  Publishing Coverage report....
10:56:48  A total of 1 reports were found
10:56:49  Code coverage enforcement failed: Method coverage in Report level 'cobertura: cobertura-coverage.xml' is lower than 60.00 healthy threshold
10:56:49  io.jenkins.plugins.coverage.exception.CoverageException: Build failed because following metrics did not meet health target: [Method {unstableThreshold=40.0, unhealthyThreshold=60.0}].
10:56:49  	at io.jenkins.plugins.coverage.CoverageProcessor.processThresholds(CoverageProcessor.java:493)
10:56:49  	at io.jenkins.plugins.coverage.CoverageProcessor.performCoverageReport(CoverageProcessor.java:125)
10:56:49  	at io.jenkins.plugins.coverage.CoveragePublisher.perform(CoveragePublisher.java:98)
10:56:49  	at jenkins.tasks.SimpleBuildStep.perform(SimpleBuildStep.java:123)
10:56:49  	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:99)
10:56:49  	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:69)
10:56:49  	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
10:56:49  	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
10:56:49  	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
10:56:49  	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
10:56:49  	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
10:56:49  	at java.lang.Thread.run(Thread.java:748)
10:56:49  
Error when executing always post condition:
java.lang.NullPointerException
	at io.jenkins.plugins.coverage.CoverageChecksPublisher.extractHealthSummary(CoverageChecksPublisher.java:172)
	at io.jenkins.plugins.coverage.CoverageChecksPublisher.extractChecksDetails(CoverageChecksPublisher.java:46)
	at io.jenkins.plugins.coverage.CoverageChecksPublisher.publishChecks(CoverageChecksPublisher.java:38)
	at io.jenkins.plugins.coverage.CoveragePublisher.perform(CoveragePublisher.java:108)
	at jenkins.tasks.SimpleBuildStep.perform(SimpleBuildStep.java:123)
	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:99)
	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:69)
	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

In the jenkins console, I was able to find what was the `CoverageAction` and it does not contain a HealthReport, causing the error on https://github.com/jenkinsci/code-coverage-api-plugin/blob/8f8d7322fda886e50c983aff1a863c4b8297187c/src/main/java/io/jenkins/plugins/coverage/CoverageChecksPublisher.java#L172

```
import io.jenkins.plugins.coverage.CoverageAction
def build = Jenkins.instance.getItemByFullName("my-build").getBuildByNumber(1)
def actions = build.getActions(CoverageAction)
actions[0].dump()

Result: <io.jenkins.plugins.coverage.CoverageAction@32ce652b owner=my-build #1 report=java.lang.ref.WeakReference@1e2f3294 healthReport=null failMessage=Build failed because following metrics did not meet health target: [Method {unstableThreshold=40.0, unhealthyThreshold=60.0}].>
```

**Comment by Cetlan** [comment]
Looks like the `processThresholds` function that's responsible for creating and returning the `HealthReport` (https://github.com/jenkinsci/code-coverage-api-plugin/blob/5c6098c4af9e242255a45bd95f8776c3b221209c/src/main/java/io/jenkins/plugins/coverage/CoverageProcessor.java#L444) is handling failUnstable/failUnhealthy by throwing a CoverageException, so the `HealthReport` never gets returns the report that's expected by the checks publisher.

I'd guess this either needs the checks publisher to not assume/require a HealthReport, or needs the failUnstable/failUnhealthy flow to be changed to not rely on exceptions. Making the HealthReport optional would let the check publish a correct status message (as the `processThresholds` method is also setting the failure message for the check), but the check would lose the health details obviously. Losing the exceptions would require passing additional data around, since unhealthy vs unstable details are lost after `processThresholds` completes currently - it would have to be preserved until further up the call chain so that the code currently catching exceptions could determine whether it should fail the build.

That's as far as I've dug into it. If I knew more about the intended design, and had time to set up the dev environment for Jenkins plugins, I might take a look at it. But maybe that at least points someone who's already set up in the right direction.

**Comment by uhafner** [comment]
This will be fixed with #265
